docs/data/tree-view/rich-tree-view/editing/EditWithIcons.tsx

```diff
@@ -0,0 +1,198 @@
+} from '@mui/x-tree-view/useTreeItem2/useTreeItem2.types';
+
+const MUI_X_PRODUCTS: TreeViewBaseItem[] = [
+  {
+    id: 'grid',
+    label: 'Data Grid',
+    children: [
+      { id: 'grid-community', label: '@mui/x-data-grid' },
+      { id: 'grid-pro', label: '@mui/x-data-grid-pro' },
+      { id: 'grid-premium', label: '@mui/x-data-grid-premium' },
+    ],
+  },
+  {
+    id: 'pickers',
+    label: 'Date and Time Pickers',
+    children: [
+      { id: 'pickers-community', label: '@mui/x-date-pickers' },
+      { id: 'pickers-pro', label: '@mui/x-date-pickers-pro' },
+    ],
+  },
+  {
+    id: 'charts',
+    label: 'Charts',
+    children: [{ id: 'charts-community', label: '@mui/x-charts' }],
+  },
+  {
+    id: 'tree-view',
+    label: 'Tree View',
+    children: [{ id: 'tree-view-community', label: '@mui/x-tree-view' }],
+  },
+];
+type CustomLabelProps = UseTreeItem2LabelSlotProps<{
+  editable: boolean;
+  editing: boolean;
+  toggleItemEditing: () => void;
+}>;
+
+function CustomLabel({
+  editing,
+  editable,
+  children,
+  toggleItemEditing,
+  ...other
+}: CustomLabelProps) {
+  return (
+    <TreeItem2Label {...other}>
+      {children}
+      {editable && (
+        <IconButton onClick={toggleItemEditing}>
+          <EditIcon />
+        </IconButton>
+      )}
+    </TreeItem2Label>
+  );
+}
+
+type CustomLabelInputProps = UseTreeItem2LabelInputSlotProps<{
+  editing: boolean;
+  handleCancelItemLabelEditing: (event: React.SyntheticEvent) => void;
+  handleSaveItemLabel: (event: React.SyntheticEvent, label: string) => void;
+  label: string;
+}>;
+
+const CustomLabelInput = React.forwardRef(function CustomLabelInput({
+  editing,
+  handleCancelItemLabelEditing,
+  handleSaveItemLabel,
+  label,
+  ...other
+}: CustomLabelInputProps) {
+  const [labelInputValue, setLabelInputValue] = React.useState(label);
+
+  if (!editing) {
+    return null;
+  }
+
+  return (
+    <React.Fragment>
+      <TreeItem2LabelInput
+        {...other}
+        label={label}
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          setLabelInputValue(event.target.value);
+        }}
+      />
+      <IconButton tabIndex={-1} onClick={handleCancelItemLabelEditing}>
+        <EditOffIcon />
+      </IconButton>
+      <IconButton
+        onClick={(event: React.MouseEvent) => {
+          handleSaveItemLabel(event, labelInputValue);
+        }}
+      >
+        <CheckIcon />
+      </IconButton>
+    </React.Fragment>
+  );
+});
+
+interface ExtendedTreeItemProps extends TreeItem2Props {
+  focusItem: (event: React.SyntheticEvent, itemId: TreeViewItemId) => void;
+  updateItemLabel: (itemId: TreeViewItemId, label: string) => void;
+}
+
+const CustomTreeItem2 = React.forwardRef(function MyTreeItem(
+  { focusItem, updateItemLabel, ...props }: ExtendedTreeItemProps,
+  ref: React.Ref<HTMLLIElement>,
+) {
+  const {
+    interactions: {
+      toggleItemEditing,
+      handleCancelItemLabelEditing,
+      handleSaveItemLabel,
```

Comment:
Good question...
We could consider dropping the interactions in favor of only using `publicAPI` methods at some point to avoid having to types of methods accessible on the item. We would need to drop `TreeItem` before to have a clean start.
I'm not entirely sure how the DX should look like at this level...